sycl/doc/design/PropagateCompilerFlagsToLinkerAndRuntime.md

```diff
@@ -0,0 +1,71 @@
+
+**NOTE**: This is not a final version. The document is still in progress.
+
+## Background
+
+When building an application with several source and object files, it should be possible to specify the optimization parameters individually for each source file/object file (for each invocation of the DPCPP compiler). The linker should pass the original optimization options (e.g. -O0 or -O2) used when building an object file to the device backend compiler (IGC compiler). This will improve the debugging experience by selectively disabling/enabling optimizations for each source file, and therefore achieving better debuggability and better performance as needed.
+
+The current behavior, is that the device backend optimization options are determined by the linker optimization options. If the -O0 option is specified for linker, the linker will pass -cl-opt-disable option to IGC for {*}all kernels{*}, essentially disabling optimizations globally. Otherwise, if the -O0 option is not specified for linker, it will not pass -cl-opt-disable option at all, therefore making the kernels mostly undebuggable, regardless of the original front-end compiler options.
+
+Here is an example that demonstrates this pain point:
+
+```
+icx -c -fsycl test1.c -o test1
+icx -c -O0 -fsycl test2.c -o test2
+icx -fsycl -o test test1.o test2.o
+```
+
+In this scenario, the fat binary is 'test' and there are no compilation flags sent across to the backend compiler. Though the user wanted to have full debuggability with test2.c module, some of the debuggablity is lost.
+
+Another scenario is shown below:
+
+```
+icpx -c -O0 -fsycl -g test.cpp -o test.o
+icpx -fsycl test.o
+```
+
+In this scenario, the fat binary is 'test' and there are no compilation flags sent across to the backend compiler. Though the user wanted to have full debuggability with test.cpp module, some of the debuggablity is lost. The user was not able to set a breakpoint inside device code.
+
+## Requirements
+
+In order to support module-level debuggability, the user will compile different module files with different levels of optimization. These optimization levels must be preserved and made use of during every stage of compilation. Following are the requirements for this feature.
+- If the user specifies '-Ox' as a front-end compile option for a particular module, this option must be preserved during compilation, linking, AOT compilation as well as JIT compilation.
+- If the user specifies '-Ox' option as a front-end linker option, this option will override any front-end compile options and the linker option will be preserved during AOT and JIT compilation.
+- If the user specifies '-O0' option, we need to pass '-cl-opt-disable' to AOT and JIT compilation stages.
```

Comment:
cl-opt-disable is an option being passed to the backend compiler. I don't think we need to consider whether it's L0 or OpenCL.

Reply:
[Level Zero](https://spec.oneapi.io/level-zero/latest/core/api.html#ze-module-desc-t) spec does not list `-cl-opt-disable` as a supported option. It is not clear to me if [`zeModuleCreate`](https://spec.oneapi.io/level-zero/latest/core/api.html#zemodulecreate) must return an error for unsupported options and which one it would be, but I would prefer if we don't pass there options which are not considered supported.